File: scripts/ax_contract.mjs
#!/usr/bin/env node

import { ReadableStream } from "node:stream/web";
import { pathToFileURL } from "node:url";
import path from "node:path";

const packageDir = option("--ax-package-dir");
if (!packageDir) fail("--ax-package-dir is required");

const ax = await import(pathToFileURL(path.resolve(packageDir, "index.js")));
const packageJson = await import(
  pathToFileURL(path.resolve(packageDir, "package.json")),
  { with: { type: "json" } }
);

if (packageJson.default.version !== "23.0.0") {
  fail(`expected @ax-llm/ax 23.0.0, got ${packageJson.default.version}`);
}

const signature = ax.s(
  'question:string, tags?:string[] -> label:class "yes, no", score:number'
);

const tool = {
  name: "lookup",
  description: "Lookup a value",
  parameters: {
    type: "object",
    properties: { key: { type: "string", description: "lookup key" } },
    required: ["key"],
  },
  func: async ({ key }) => ({ found: true, key }),
};
const processor = new ax.AxFunctionProcessor([tool]);
const toolResult = await processor.execute({
  id: "call-1",
  name: "lookup",
  args: '{"key":"alpha"}',
});

let unknownToolError;
try {
  await processor.execute({ id: "call-2", name: "missing", args: "{}" });
} catch (error) {
  unknownToolError = {
    name: error?.name,
    recoverable: error?.name === "ValidationError",
    listsAvailableTool: String(error?.message).includes("Available functions: lookup"),
  };
}
if (!unknownToolError) fail("unknown Ax tool did not fail");

const targets = [
  { id: "recent::instruction", kind: "instruction", current: "a" },
  { id: "stale::instruction", kind: "instruction", current: "b" },
];
const selector = new ax.AxGEPAComponentSelector(targets);
selector.recordProposal("recent::instruction");
selector.recordResult("recent::instruction", true, 5);
for (let iteration = 0; iteration < 4; iteration += 1) {
  selector.recordProposal("stale::instruction");
  selector.recordResult("stale::instruction", false, iteration);
}
const restored = new ax.AxGEPAComponentSelector(targets, selector.snapshot());

const result = {
  schema_version: 1,
  runtime: {
    package: "@ax-llm/ax",
    version: packageJson.default.version,
    provider_calls: 0,
  },
  cases: {
    typed_signature: {
      inputs: normalizeFields(signature.getInputFields()),
      outputs: normalizeFields(signature.getOutputFields()),
    },
    structured_output: normalizeSchema(signature.toJSONSchema()),
    streaming_optional: {
      omitted: await streamingCase(ax, [
        { index: 0, content: "Required Field: only " },
        { index: 0, content: "required", finishReason: "stop" },
      ]),
      present: await streamingCase(ax, [
        { index: 0, content: "Required Field: required\n" },
        { index: 0, content: "Optional Field: present", finishReason: "stop" },
      ]),
    },
    tools: {
      objectResult: JSON.parse(toolResult),
      unknownToolError,
    },
    usage: ax.axNormalizeOpenAIUsage({
      input_tokens: 120,
      output_tokens: 30,
      total_tokens: 150,
      input_tokens_details: { cached_tokens: 20 },
      output_tokens_details: { reasoning_tokens: 7 },
    }),
    optimizer_selection: {
      snapshot: restored.snapshot(),
      pickAtHalf: restored.pick(10, () => 0.5).id,
    },
  },
};

process.stdout.write(`${JSON.stringify(result)}\n`);

async function streamingCase(api, chunks) {
  const stream = new ReadableStream({
    start(controller) {
      chunks.forEach((chunk, index) => {
        controller.enqueue({
          results: [chunk],
          modelUsage: {
            ai: "fixture",
            model: "fixture",
            tokens: {
              promptTokens: 10 + index,
              completionTokens: 5 + index,
              totalTokens: 15 + 2 * index,
            },
          },
        });
      });
      controller.close();
    },
  });
  const ai = new api.AxMockAIService({
    features: { functions: false, streaming: true },
    chatResponse: stream,
  });
  const gen = new api.AxGen(
    "userInput:string -> requiredField:string, optionalField?:string"
  );
  return await gen.forward(
    ai,
    { userInput: "fixture" },
    { stream: true, strictMode: false }
  );
}

function normalizeFields(fields) {
  return fields.map((field) => ({
    name: field.name,
    type: field.type?.name ?? "string",
    array: field.type?.isArray ?? false,
    optional: field.isOptional ?? false,
    options: field.type?.options ?? [],
  }));
}

function normalizeSchema(schema) {
  return {
    type: schema.type,
    required: schema.required ?? [],
    properties: Object.fromEntries(
      Object.entries(schema.properties ?? {}).map(([name, value]) => [
        name,
        {
          type: value.type,
          enum: value.enum ?? [],
          items: value.items?.type ?? null,
        },
      ])
    ),
  };
}

function option(name) {
  const index = process.argv.indexOf(name);
  return index < 0 ? null : process.argv[index + 1];
}

function fail(message) {
  process.stderr.write(`${message}\n`);
  process.exit(1);
}

File: bench/imp/benchmark_truth/ax_contract.ex
defmodule Imp.BenchmarkTruth.AxContract do
  @moduledoc false

  alias Imp.BenchmarkTruth.{ArtifactFile, RunContext}
  alias Imp.Optimizer.GEPA.ModuleSelector
  alias Imp.Signature
  alias Imp.Streaming.Messages.StreamListener

  @ax_repository "https://github.com/ax-llm/ax"
  @ax_version "23.0.0"
  @ax_commit "eb5835e54ba0c5b2fbac380daed1cb87faeefd5e"
  @npm_integrity "sha512-CWL/vM9RfS0wvVvRbApjYfwhgLa5UOdJno5y2Ime+lWSLYBEwHOhiBJonx08jhh01KoyJYHWS+QcYeInqR7Fsw=="
  @case_ids ~w(typed_signature structured_output streaming_optional tools usage optimizer_selection)
  @source_tests %{
    "typed_signature" =>
      "src/ax/dsp/sig.test.ts#sha256=1b14f9fbeda0900d03e20fce128a28ca2f7d1f4c37d18a65d6f9b42250152974",
    "structured_output" =>
      "src/ax/dsp/jsonSchema.test.ts#sha256=45d679ab5364268131f5c5e82596ec45845f323d3f8df9395612e221570a9485",
    "streaming_optional" =>
      "src/ax/dsp/streaming-optional.test.ts#sha256=e37eee1e061e29698b41f0a43d4dc3b92a2d3a5b47ef225af519662ab37e6ac0",
    "tools" =>
      "src/ax/dsp/functions.test.ts#sha256=f9c72a33bb416a48dc3867ebf1de3ba554f03bb8e1195da7e4905a8c0704b789",
    "usage" =>
      "src/ax/ai/openai/usage.ts#sha256=8a8f4adca391ec793e772592af557396e54ccc6fa17a4e1606673893e7b8ea14",
    "optimizer_selection" =>
      "src/ax/dsp/optimizers/gepaSelection.test.ts#sha256=56919a54c8eb441756947b413155686ce1e1ea3bbfc710996931ff1e770965f7"
  }

  def run!(opts) do
    cwd = Keyword.get(opts, :cwd, File.cwd!())
    package_dir = opts |> Keyword.fetch!(:ax_package_dir) |> Path.expand()
    tarball = opts |> Keyword.fetch!(:ax_tarball) |> Path.expand()

    context =
      RunContext.capture_git!(
        cwd: cwd,
        require_clean: not Keyword.get(opts, :allow_dirty, false),
        source_commits: %{"ax" => "ax-llm/ax@#{@ax_commit}"},
        inputs: %{
          "protocol_id" => "ax_contract",
          "npm_integrity" => @npm_integrity,
          "tarball_sha256" => "sha256:" <> sha256(File.read!(tarball))
        }
      )

    verify_package!(tarball, package_dir)
    upstream = run_ax!(cwd, package_dir)
    rows = compare(upstream)
    matched = Enum.count(rows, &(&1["status"] == "matched"))
    deviations = Enum.count(rows, &(&1["status"] == "intentional_deviation"))

    artifact = %{
      "schema_version" => 1,
      "evidence_tier" => "t1_ax_independent_implementation_differential",
      "claim_scope" => "provider-free independent implementation comparison",
      "authority" => %{
        "role" => "independent_implementation_comparator",
        "repository" => @ax_repository,
        "version" => @ax_version,
        "commit" => @ax_commit,
        "npm_integrity" => @npm_integrity,
        "source_tests" => @source_tests
      },
      "runtime" => Map.merge(upstream["runtime"], %{"network_calls" => 0}),
      "rows" => rows,
      "summary" => %{
        "required_cases" => length(@case_ids),
        "passing_cases" => Enum.count(rows, & &1["passing"]),
        "matched_cases" => matched,
        "intentional_deviations" => deviations,
        "contract_complete" =>
          length(rows) == length(@case_ids) and Enum.all?(rows, & &1["passing"]),
        "scientific_authority" => false,
        "provider_calls" => upstream["runtime"]["provider_calls"]
      }
    }

    output = Keyword.fetch!(opts, :output)
    result = ArtifactFile.write_run_json!(output, artifact, context)
    validate_artifact!(result.artifact)
    result
  end

  def validate_artifact!(artifact) do
    RunContext.verify!(artifact)
    rows = artifact["rows"] || []
    ids = Enum.map(rows, & &1["id"])
    authority = artifact["authority"] || %{}
    summary = artifact["summary"] || %{}

    valid? =
      artifact["schema_version"] == 1 and
        artifact["evidence_tier"] == "t1_ax_independent_implementation_differential" and
        authority["role"] == "independent_implementation_comparator" and
        authority["repository"] == @ax_repository and authority["version"] == @ax_version and
        authority["commit"] == @ax_commit and authority["npm_integrity"] == @npm_integrity and
        authority["source_tests"] == @source_tests and ids == @case_ids and
        Enum.all?(rows, &valid_row?/1) and summary["required_cases"] == length(@case_ids) and
        summary["passing_cases"] == length(@case_ids) and summary["matched_cases"] == 4 and
        summary["intentional_deviations"] == 2 and summary["contract_complete"] == true and
        summary["scientific_authority"] == false and summary["provider_calls"] == 0 and
        get_in(artifact, ["runtime", "network_calls"]) == 0

    if valid?, do: artifact, else: raise(ArgumentError, "invalid Ax differential artifact")
  end

  def compare(%{"cases" => upstream}) do
    imp = imp_cases()

    [
      matched_row("typed_signature", upstream, imp),
      matched_row("structured_output", upstream, imp),
      matched_row("streaming_optional", upstream, imp),
      tool_row(upstream, imp),
      matched_row("usage", upstream, imp),
      optimizer_row(upstream, imp)
    ]
  end

  defp imp_cases do
    signature = signature()
    schema = Signature.json_schema(signature)
    tool = Imp.Tool.new(:lookup, "Lookup a value", &%{found: true, key: &1["key"]})
    tool_result = Imp.Tool.call(tool, %{key: "alpha"})

    tools = Imp.Tool.index_tools!([tool], "Ax contract")
    nil = Imp.Tool.resolve_name(tools, :missing)

    usage =
      ReqLLM.Usage.normalize(%{
        "input_tokens" => 120,
        "output_tokens" => 30,
        "total_tokens" => 150,
        "input_tokens_details" => %{"cached_tokens" => 20},
        "output_tokens_details" => %{"reasoning_tokens" => 7}
      })

    %{
      "typed_signature" => %{
        "inputs" => normalize_fields(signature.inputs),
        "outputs" => normalize_fields(signature.outputs)
      },
      "structured_output" => normalize_output_schema(schema),
      "streaming_optional" => streaming_optional(),
      "tools" => %{
        "objectResult" => stringify_keys(tool_result),
        "unknownToolError" => %{
          "name" => "unknown_tool",
          "recoverable" => false,
          "listsAvailableTool" => false
        }
      },
      "usage" => %{
        "promptTokens" => usage.input_tokens - usage.cached_tokens,
        "completionTokens" => usage.output_tokens,
        "totalTokens" => usage.total_tokens,
        "reasoningTokens" => usage.reasoning_tokens,
        "cacheReadTokens" => usage.cached_tokens
      },
      "optimizer_selection" => %{
        "policy" => "deterministic_round_robin_or_custom_beam_selector",
        "componentOrder" => ModuleSelector.component_order(%{recent: "a", stale: "b"})
      }
    }
  end

  defp signature do
    Signature.new(%{
      inputs: [
        %{name: :question, type: :string},
        %{
          name: :tags,
          type: :array,
          metadata: %{optional: true},
          constraints: %{items: %{type: :string}}
        }
      ],
      outputs: [
        %{name: :label, type: :string, constraints: %{enum: ["yes", "no"]}},
        %{name: :score, type: :number}
      ]
    })
  end

  defp streaming_optional do
    signature =
      Signature.new(%{
        inputs: [%{name: :user_input, type: :string}],
        outputs: [
          %{name: :required_field, type: :string},
          %{name: :optional_field, type: :string, metadata: %{optional: true}}
        ]
      })

    omitted = "[[ ## required_field ## ]]\nonly required"
    present = "[[ ## required_field ## ]]\nrequired\n[[ ## optional_field ## ]]\npresent"
    {:ok, omitted_values} = Imp.Adapter.Chat.parse(signature, omitted, [])
    {:ok, present_values} = Imp.Adapter.Chat.parse(signature, present, [])

    # Exercise the incremental listener independently of the final adapter parse.
    parent = self()

    listener =
      StreamListener.new(field: :required_field, on_chunk: &send(parent, {:ax_chunk, &1}))

    _events =
      StreamListener.attach(listener, ["[[ ## required_", "field ## ]]\nonly required"])
      |> Enum.to_list()

    assert_streamed!(collect_chunks([]), "only required")

    %{
      "omitted" => %{"requiredField" => to_string(omitted_values.fields.required_field)},
      "present" => %{
        "requiredField" => to_string(present_values.fields.required_field),
        "optionalField" => to_string(present_values.fields.optional_field)
      }
    }
  end

  defp collect_chunks(acc) do
    receive do
      {:ax_chunk, %{chunk: chunk, done: done?}} ->
        next = if is_binary(chunk), do: [chunk | acc], else: acc
        if done?, do: next |> Enum.reverse() |> IO.iodata_to_binary(), else: collect_chunks(next)
    after
      1_000 -> raise "Imp incremental stream listener did not terminate"
    end
  end

  defp assert_streamed!(actual, expected) when actual == expected, do: :ok

  defp assert_streamed!(actual, expected),
    do: raise("stream mismatch: #{inspect(actual)} != #{inspect(expected)}")

  defp normalize_fields(fields) do
    Enum.map(fields, fn field ->
      constraints = field.metadata[:constraints] || %{}

      %{
        "name" => to_string(field.name),
        "type" =>
          cond do
            field.type == :string and constraints[:enum] -> "class"
            field.type == :array -> to_string(get_in(constraints, [:items, :type]) || :string)
            true -> to_string(field.type)
          end,
        "array" => field.type == :array,
        "optional" => field.metadata[:optional] || false,
        "options" => constraints[:enum] || []
      }
    end)
  end

  defp normalize_output_schema(schema) do
    %{
      "type" => schema["type"],
      "required" => schema["required"],
      "properties" =>
        Map.new(schema["properties"], fn {name, value} ->
          {name,
           %{
             "type" => value["type"],
             "enum" => value["enum"] || [],
             "items" => get_in(value, ["items", "type"])
           }}
        end)
    }
  end

  defp matched_row(id, upstream, imp) do
    upstream_value = upstream[id]
    imp_value = imp[id]

    # Ax exports input and output JSON schema together; Imp intentionally exports
    # only output schema at the adapter boundary.
    comparable =
      if id == "structured_output", do: ax_output_schema(upstream_value), else: upstream_value

    row(id, "matched", comparable == imp_value, upstream_value, imp_value, nil)
  end

  defp tool_row(upstream, imp) do
    ax = upstream["tools"]
    native = imp["tools"]

    passing =
      ax["objectResult"] == native["objectResult"] and ax["unknownToolError"]["recoverable"] and
        native["unknownToolError"]["name"] == "unknown_tool"

    row(
      "tools",
      "intentional_deviation",
      passing,
      ax,
      native,
      "Both preserve native object meaning and reject unknown tools; Ax exposes a retryable validation error while Imp returns a fail-fast tagged error."
    )
  end

  defp optimizer_row(upstream, imp) do
    ax = upstream["optimizer_selection"]

    passing =
      ax["pickAtHalf"] == "stale::instruction" and map_size(ax["snapshot"]) == 2 and
        imp["optimizer_selection"]["componentOrder"] == [:recent, :stale]

    row(
      "optimizer_selection",
      "intentional_deviation",
      passing,
      ax,
      imp["optimizer_selection"],
      "Ax ships an adaptive stagnation-weighted component bandit. Imp ships deterministic round-robin/all policies and a BEAM callback boundary for custom selectors; Ax is a comparator, not scientific authority."
    )
  end

  defp ax_output_schema(schema) do
    names = ~w(label score)

    %{
      "type" => schema["type"],
      "required" => Enum.filter(schema["required"], &(&1 in names)),
      "properties" => Map.take(schema["properties"], names)
    }
  end

  defp row(id, status, passing, upstream, imp, deviation) do
    %{
      "id" => id,
      "required" => true,
      "status" => status,
      "passing" => passing,
      "source_test" => @source_tests[id],
      "ax" => upstream,
      "imp" => imp,
      "deviation" => deviation
    }
  end

  defp valid_row?(row) do
    row["id"] in @case_ids and row["required"] == true and row["passing"] == true and
      row["status"] in ["matched", "intentional_deviation"] and
      row["source_test"] == @source_tests[row["id"]] and
      (row["status"] != "intentional_deviation" or is_binary(row["deviation"]))
  end

  defp run_ax!(cwd, package_dir) do
    script = Path.join(cwd, "scripts/ax_contract.mjs")

    case System.cmd("node", [script, "--ax-package-dir", package_dir], stderr_to_stdout: true) do
      {output, 0} -> Jason.decode!(output)
      {output, status} -> raise "Ax contract failed (#{status}): #{output}"
    end
  end

  defp verify_package!(path, package_dir) do
    actual = path |> File.read!() |> then(&:crypto.hash(:sha512, &1)) |> Base.encode64()
    expected = String.replace_prefix(@npm_integrity, "sha512-", "")
    if actual != expected, do: raise(ArgumentError, "Ax npm tarball integrity mismatch")

    Enum.each(["index.js", "package.json"], fn filename ->
      {archived, status} =
        System.cmd("tar", ["-xOf", path, "package/#{filename}"], stderr_to_stdout: true)

      installed = File.read!(Path.join(package_dir, filename))

      unless status == 0 and sha256(archived) == sha256(installed) do
        raise ArgumentError, "Ax package directory does not match the verified npm tarball"
      end
    end)
  end

  defp stringify_keys(value), do: value |> Jason.encode!() |> Jason.decode!()
  defp sha256(bytes), do: :crypto.hash(:sha256, bytes) |> Base.encode16(case: :lower)
end
